### PR TITLE
Add support for RVV f32-qs8/qu8-vcvt microkernels, test and config.

### PR DESCRIPTION
**Pull request by KaustubhIMG**

### CODE REVIEW

**Comment by oliIMG** [comment]
Related to:
https://github.com/google/XNNPACK/pull/6908

**Comment by KaustubhIMG** [comment]
@fbarchard @allenkelly @bhbruce @keeranroth, If you have time, please help review this PR.

File: src/f32-qs8-vcvt/rvv.c.in [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include <riscv_vector.h>
+#include "xnnpack/math.h"
+#include "xnnpack/vcvt.h"
+
+
+void xnn_f32_${DATATYPE.lower()}_vcvt_ukernel__rvv_u${LMUL}v(
+    size_t batch,
+    const float* input,
+    ${XINT8_T}* output,
+    const union xnn_f32_${DATATYPE.lower()}_cvt_params params[restrict XNN_MIN_ELEMENTS(1)]) XNN_OOB_READS
```
Comment (fbarchard) [inline src/f32-qs8-vcvt/rvv.c.in:21]: RVV never has OOB_READS?   this tells asan not to check reads, but we have vsetvl that should ensure reads stay in bounds

File: src/f32-qs8-vcvt/rvv.c.in [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  assert(input != NULL);
+  assert(output != NULL);
+
+  batch >>= XNN_LOG2_SIZEOF_FLOAT;
+
+  const float scale = params->scalar.scale;
+  const vfloat32m${LMUL}_t magic_bias_f32v = __riscv_vfmv_v_f_f32m${LMUL}(12582912.0f, __riscv_vsetvl_e32m${LMUL}(batch));
+  const float output_min_less_zero_point = (float) ((int32_t) params->scalar.output_min - (int32_t) params->scalar.output_zero_point);
+  const float output_max_less_zero_point = (float) ((int32_t) params->scalar.output_max - (int32_t) params->scalar.output_zero_point);
+  const int32_t magic_min = (int32_t) float_as_uint32(12582912.0f + output_min_less_zero_point);
```
Comment (fbarchard) [inline src/f32-qs8-vcvt/rvv.c.in:34]: is the 'magic' method really the fastest on riscv?

The basic function would load floats, scale, bias, clamp and convert to bytes.
the clamping can be done as bytes if it is faster.

Also I think it is common to not require clamping, so we may want to consider a 'linear' version.
In the past we would also consider doing multiple versions - e.g. fmagic as a 'variant' or different template, incase some hardware prefers one way or the other.

**Review by fbarchard — APPROVED** [review]
note the qu8 is not used much... we are trying to move away from qu8 and use qs8 and qd8 for gemms.
f32-qs8-vcvt is used alot.  we may want f16-qs8-vcvt too

there are global changes to do with params and tests that may require rebasing

File: src/f32-qs8-vcvt/rvv.c.in (hunk 2 above)
Comment (oliIMG) [reply src/f32-qs8-vcvt/rvv.c.in:34]: The magic implementation saves one vector instruction, and doesn't utilize as many converts, if you prefer one variant over the other let me know which you would like me to go forward with. I can do multiple versions if you deem it necessary also.

File: src/f32-qs8-vcvt/rvv.c.in (hunk 1 above)
Comment (oliIMG) [reply src/f32-qs8-vcvt/rvv.c.in:21]: Other RVV kernels had this enabled by default so I didn't investigate further into this. Would we want to be consistent with other kernels, or remove this?

**Comment by oliIMG** [comment]
> note the qu8 is not used much... we are trying to move away from qu8 and use qs8 and qd8 for gemms. f32-qs8-vcvt is used alot. we may want f16-qs8-vcvt too
> 
> there are global changes to do with params and tests that may require rebasing

Currently no patch is ready for f16-qs8, will consider this in the future. 
Also regarding qu8, it's not much more work to add support for this, but would you like to no longer include qu8 kernels going forward?

File: cmake/gen/rvv_microkernels.cmake [hunk 3]
```diff
@@ -72,6 +74,12 @@ SET(NON_PROD_RVV_MICROKERNEL_SRCS
   src/f32-igemm/gen/f32-igemm-7x4v-relu-rvv.c
   src/f32-igemm/gen/f32-igemm-7x4v-rvv.c
   src/f32-maxpool/gen/f32-maxpool-9p8x-minmax-rvv-c1v.c
+  src/f32-qs8-vcvt/gen/f32-qs8-vcvt-rvv-u1v.c
+  src/f32-qs8-vcvt/gen/f32-qs8-vcvt-rvv-u4v.c
+  src/f32-qs8-vcvt/gen/f32-qs8-vcvt-rvv-u8v.c
+  src/f32-qu8-vcvt/gen/f32-qu8-vcvt-rvv-u1v.c
+  src/f32-qu8-vcvt/gen/f32-qu8-vcvt-rvv-u4v.c
+  src/f32-qu8-vcvt/gen/f32-qu8-vcvt-rvv-u8v.c
   src/f32-raddstoreexpminusmax/gen/f32-raddstoreexpminusmax-rvv-rr2-p6-u2v.c
```
Comment (bhbruce) [inline cmake/gen/rvv_microkernels.cmake:83]: Do you use the script to generate this? It seems that u2 is disappeared.

Comment (oliIMG) [reply cmake/gen/rvv_microkernels.cmake:83]: Yes utilized tools/update-microkernels, it's in listed in the PROD_SRC list, should it also be in the the NON_PROD list?

File: src/configs/unary-elementwise-config.c [hunk 4]
```diff
@@ -1627,10 +1627,11 @@ static void init_f32_to_qs8_cvt_config(void) {
       f32_to_qs8_cvt_config.init.f32_qs8_cvt = xnn_init_f32_qs8_cvt_scalar_params;
       f32_to_qs8_cvt_config.element_tile = 4;
     }
-  #elif XNN_ARCH_RISCV
-    f32_to_qs8_cvt_config.ukernel = (xnn_vunary_ukernel_fn) xnn_f32_qs8_vcvt_ukernel__scalar_lrintf_u4;
+  #elif XNN_ARCH_RISCV && XNN_ENABLE_RISCV_VECTOR
+    const struct xnn_hardware_config* hardware_config = xnn_init_hardware_config();
+    f32_to_qs8_cvt_config.ukernel = (xnn_vunary_ukernel_fn) xnn_f32_qs8_vcvt_ukernel__rvv_u2v;
```
[the remark this answers is not in the document]
Comment (oliIMG) [reply src/configs/unary-elementwise-config.c:1632]: You left a comment on the previous PR regarding this, and said to not utilize either u1v or u8v, I can bump it up to u4v or u8v if necessary.

File: cmake/gen/rvv_microkernels.cmake (hunk 3 above)
Comment (bhbruce) [reply cmake/gen/rvv_microkernels.cmake:83]: Got it. Thx.

**Comment by dsharlet** [comment]
This branch needs to merge master in order to merge. There are no conflicts, but there are breaking changes: the params type is now a struct and not a union.

**Review by dsharlet — APPROVED** [review]

**Comment by KaustubhIMG** [comment]
Already merged

**Comment by dsharlet** [comment]
There are some new flaky tests that seem related to this PR: https://github.com/google/XNNPACK/actions/runs/10857812047/job/30134876420?pr=7089

Can you please take a look?

**Comment by KaustubhIMG** [comment]
@dsharlet Added fix in https://github.com/google/XNNPACK/pull/7096

**Pull request closed without merging**